semgrep-core/src/analyzing/Dataflow.ml

```diff
@@ -264,10 +267,24 @@ module Make (F : Flow) = struct
       fixpoint_worker eq mapping trans flow succs work''
 
   let forward_succs (f : F.flow) n =
-    (f#successors n)#fold (fun s (ni, _) -> NodeiSet.add ni s) NodeiSet.empty
+    (f.graph#successors n)#fold
+      (fun s (ni, _) -> NodeiSet.add ni s)
+      NodeiSet.empty
 
   let backward_succs (f : F.flow) n =
-    (f#predecessors n)#fold (fun s (ni, _) -> NodeiSet.add ni s) NodeiSet.empty
+    (f.graph#predecessors n)#fold
+      (fun s (ni, _) -> NodeiSet.add ni s)
+      NodeiSet.empty
+
+  let mk_worklist ({ graph; entry } : F.flow) =
+    let rec aux nodei seen =
+      if NodeiSet.mem nodei seen then seen
+      else
+        let seen = NodeiSet.add nodei seen in
+        let succs = forward_succs { graph; entry } nodei in
+        NodeiSet.fold aux succs seen
```

Comment:
Correct me if I'm wrong, but here you may be visiting the same nodes several times. The `seen` set only knows about the nodes visited in one specific branch of the traversal.